[AIP-49] Add Otel Framework and implement Counters

ferruzzi opened the pull request:
AIP-49: Implement OpenTelemetry

**Overview**
Adds the support framework for OpenTelemetry support and implements Counters and UpDownCounters.  Gauges and Timers build on this and are Coming Soon :tm:

**Testing**
Passes `breeze static-checks --all-files`, `breeze build-docs`, and `breeze testing tests --run-in-parallel` locally.

cc: @o-nikolas @vincbeck @vandonr-amz @syedahsn @howardyoo @potiuk

ferruzzi committed:
Add Otel Framework

ferruzzi committed:
fix off-by-one error

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        # TODO: I don't think this is the right use for rate???
+        value = count * rate
```
For discussion:   I'm not sure how we should handle rate here.   [The StatsD implementation](https://github.com/jsocol/pystatsd/blob/main/statsd/client/base.py#L63) picked a random number and if it was less than `rate` then it did nothing, I think?  Is that the expected behavior or is there more there that I am missing?  Should the value be increased accordingly or it it purely a throttle control?

((cc @howardyoo ))

o-nikolas replied:
> I'm not sure how we should handle rate here. [The StatsD implementation](https://github.com/jsocol/pystatsd/blob/main/statsd/client/base.py#L63) picked a random number and if it was less than rate then it did nothing, I think?

It looks like this is used as a rate-limiting feature. You provide a rate which must be between 0.00 and 0.99 (basically a percent). Then a random number between 0.00 and 1.00 is generated and if it's larger than the rate value you provided it short circuits and if it's less, then it emits the metric.
So if you provide 0.95 as your rate, most numbers generated by `random.random()` will be below 0.95 so you'll emit most of the time (i.e. emit 95% of the time). If you provide a rate of 0.05, most numbers generated by `random.random()` will be above that and so it'll short circuit and not emit the metric (i.e. only emit 5% of the time).
If you provide a value greater than or equal to 1 the whole rate limiting branch is not run (because you're saying you want it emitted 100% of the time essentially)

ferruzzi replied:
In https://github.com/apache/airflow/pull/30873/commits/102245027f4bbdc7459b53fc90bb08686e38873f I've implemented it solely as a flood control, no value scaling.

o-nikolas commented on airflow/jobs/local_task_job_runner.py:
```diff
@@ -306,3 +306,13 @@ def _log_return_code_metric(self, return_code: int):
             "local_task_job.task_exit."
             f"{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}"
         )
+        # Same metric with tagging
+        Stats.incr(
```
It's unfortunate that you can't detect these cases inside `incr(...)` :cry:

ferruzzi replied:
That was one of the ideas I had for the Metrics Registry, actually.   It could be used to record the template used to generate the old names and that template could be used to extract it down into the new names.  Unfortunately there isn't a great deal of consistency on the names at this point so it's not as easy as just doing a split(), sadly.

ferruzzi replied:
Or maybe it can... if i pass in the new name with the attributes/tags, then in incr() just emit it that way and then emit it concatenated all together... but then I'd have to worry about sorting order of the items in `attributes`which might be a different nightmare...

Maybe adjust incr() so it takes an optional "back-compat-name" parameter?  If it exists, then emit both?

o-nikolas replied:
> Maybe adjust incr() so it takes an optional "back-compat-name" parameter? If it exists, then emit both?

Then it's up to the caller to know the right name and character limit right?

ferruzzi replied:
I just saw that last comment.   Yeah, it would be up to whoever is adding a new metric to know about the character limit on the name.  Maybe have a look at the latest implementation of incr() and let me know what you think.  If we go with the current technique, then we can get rid of these double calls all over the place and consolidate them to `incr("new_name, tags=tags, back_compat_name=f'name_{with}_{embedded}_{tags}`

ferruzzi replied:
If/When we move over to the centralized Registry idea, then a pre-commit could enforce it there

ferruzzi replied:
Had a chat with Niko and we've decided the redundant calls here are the lesser of the two evils and will make it easier to deprecate the long names at a later date rather than hiding the problem. I'm going to roll back that change in a coming update to this PR and leave all the duplicate calls in place.  Resolving.

o-nikolas commented on airflow/jobs/local_task_job_runner.py:
```diff
@@ -306,3 +306,13 @@ def _log_return_code_metric(self, return_code: int):
             "local_task_job.task_exit."
             f"{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}"
         )
+        # Same metric with tagging
```
Maybe add a bit about why you're doing this? Can link to an issue or discussion as well.

ferruzzi replied:
See below, I may have just come up with a slightly more elegant solution [here](https://github.com/apache/airflow/pull/30873/files#r1178495728)

ferruzzi replied:
Had a chat with Niko offline and we've decided the redundant calls here are the lesser of the two evils and will make it easier to deprecate the long names at a later date rather than hiding the problem.  I'm going to roll back that change in a coming update to this PR and leave all the duplicate calls in place.

howardyoo replied:
@ferruzzi , BTW, I recently met this particular metric to be somewhat of a problem where when exported to datadog, these long metric names were introducing too high of the cardinality, FYI, had to ask around what were the usage of this metric and, somebody explained that it was an instrumentation that somebody made which kept track of whether the local task job run did end successfully or hung up (or exited with error). This metric would only be for local task job runs, and therefore, I wonder if this metric is actually that much useful at all or not, since I do feel like this is not actually a telemetry data (e.g. metric), but more so like an event.

ferruzzi replied:
Can you start a ticket/discussion for this?  I know a few other folks have mentioned certain stats that either don't work or don't serve the purpose they claim to.  Let's get a discussion going and collect those in one place.  Once this OTel support is in, we can go through that list and clean stuff up.  At that point we'll already be deprecating and refactoring metrics stuff and will likely need a major release, so it seems like a good time to go though with those fixes/changes as well.

ferruzzi replied:
Current behavior [here](https://github.com/apache/airflow/blob/main/airflow/metrics/validators.py#L53) is to catch it, log it, and move on.

howardyoo replied:
> Can you start a ticket/discussion for this? I know a few other folks have mentioned certain stats that either don't work or don't serve the purpose they claim to. Let's get a discussion going and collect those in one place. Once this OTel support is in, we can go through that list and clean stuff up. At that point we'll already be deprecating and refactoring metrics stuff and will likely need a major release, so it seems like a good time to go though with those fixes/changes as well.

Sure, will do.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+
+
+class SafeOtelLogger:
+    """Otel Logger"""
```
Nit: Maybe add some more useful info or just drop this

ferruzzi replied:
Totally fair.  StatsD Logger and DogStatsD Logger had that so i just followed suit blindly, but it's pretty useless.

o-nikolas replied:
Bump, this thread was resolved but the comment still looks the same

ferruzzi replied:
I tried to drop it my MyPy won't let me and I really don't know what else to say there.   I could expand it to something like "This class contains the logic to implement the OpenTelemetry Protocol" but is anything really gained by that?

o-nikolas replied:
Whoa, I swear I've seen lots of classes that don't have doc strings, interesting. Okay, don't worry about it then.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+from airflow.metrics.protocols import DeltaType, Timer, TimerProtocol
+from airflow.metrics.validators import AllowListValidator, validate_stat
+
+# This is currently the only UDC used.  If more are added, we should add a better system for this.
```
Can you link to what an updown counter is and more context on why this is the only one and why it's important to treat it differently?

ferruzzi replied:
Generally a counter is monotonic (can only go up) but OpenTelemetry has an UpDownCounter which, I'm sure you can guess, is non-monotonic; it can go up or down.  The choice here is to either drop this metric and implement the rest as monotonic Counters, implement all counters as UpDownCounters, or add a bit of logic to do it intelligently.  The catch is that the Collector which transmits these metrics to the upstream dashboard tools (Prometheus, Grafana, etc) assigns the type of `gauge` to any UDC instead of Counter.  Adding this logic feels like the best solution so normal counters still get `typed` correctly and we don't lose an existing metric.

OTel docs for [Counters](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#counter-creation) and [UDC](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter)

o-nikolas replied:
Yupp, all that makes sense, I just want more of that context in the comment :)

ferruzzi replied:
Added in https://github.com/apache/airflow/pull/30873/commits/f90ad01543c326078d7a6bc8b385029768f31c97

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        Decrement stat by count.
+
+        :param stat: The name of the stat to increment.
```
s/increment/decrement/g

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        # TODO: I don't think this is the right use for rate???
```
Agreed, see my reply to the other thread.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+
+    def get_counter(self, name: str, attributes: Attributes = None):
+        """
+        Returns the value of the counter; creates a new one if it does not exist.
```
Does it return the value of the counter? Or just the counter object itself?

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        else:
+            counter = self.meter.create_counter(name=otel_safe_name)
+
+        print(f"--> created {otel_safe_name} as type: {str(type(counter)).split('.')[-1][:-2]}")
```
There are a few of these prints which were for my own debugging use that I missed removing.  I can delete them when I make other revisions or I could leave it if it's useful for logging purposes.

o-nikolas replied:
You should be able to instantiate a logger in this module, that'd be preferable over using `print`

ferruzzi replied:
Yeah, I think I'll leave it in as a logging.info() unless someone objects.

ferruzzi replied:
Made them log messages in https://github.com/apache/airflow/pull/30873/commits/683b2bc77a4d6ff5fc9d9dd6a8c300a973582616

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+    host = conf.get("metrics", "otel_host")  # ex: "breeze-otel-collector"
+    port = conf.getint("metrics", "otel_port")  # ex: 4318
+    prefix = conf.get("metrics", "otel_prefix")  # ex: "airflow"
+    interval = conf.getint("metrics", "otel_interval_milliseconds")  # ex: 30000
```
Initially I had planned to wrap this block in a try/except that would raise a more useful error message, but in the end I decided it wasn't done elsewhere and likely not worthwhile.  Thoughts?

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        )
+    ]
+
+    # TODO:  remove this console exporter before launch??  It's handy for debugging maybe?
```
Happy to hear opinions on this.  ConsoleMetricExporter prints every metric to the terminal instead of (or in addition to ) emitting them out to a Collector.  I found it very handy for debugging but likely less useful once this is all set up and working.  I intend to remove this, but open to leaving it and plumbing that debug flag through if others think it's worth keeping handy.

o-nikolas replied:
If you are going to keep it, you should add a config option for it so that it can be enabled without modifying source code (like changing log level). But it's also fine to drop it, since it likely won't be used much.

ferruzzi replied:
Added in https://github.com/apache/airflow/pull/30873/commits/9bf93d3b08de83c253b317142e746f7b5e469cca

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        self.name: str = name
+        self.unit: str = unit
+        self.description: str = description
+        self.attributes: Attributes = attributes
```
I just realized this slipped in.  BaseInstrument is not used here, but is needed for implementing gauges which I am working on in parallel and will finalize next.  I can remove the class from this PR and include it in the next one if preferred.

ferruzzi replied:
And I've fixed the typo in the docstring :P

o-nikolas replied:
Thanks for the clarification, I did have a comment about that (which I deleted, I'll reply here). I'm happy either way, it's a small piece of code so it's not a big deal one way or another.

ferruzzi replied:
I'm doing revisions anyway, I may as well remove it here and include it later to keep the PR cleaner.  No biggie either way to me.

ferruzzi replied:
Removed in https://github.com/apache/airflow/pull/30873/commits/683b2bc77a4d6ff5fc9d9dd6a8c300a973582616

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+    allow_list_validator = AllowListValidator(allow_list)
+
+    resource = Resource(attributes={SERVICE_NAME: "Airflow"})
+    # TODO:  figure out https instead of http ??
```
This doesn't appear to be a simple task.  I've been treating it as a "nice to have" just on principle, but can I bump up the urgency on that if we think it's critical.

o-nikolas replied:
I would think most enterprise customers would want this. I don't think it blocks this PR, but we should follow soon?

ferruzzi replied:
Yeah, I agree. Initial research seems to indicate it's not an option, but I find that very hard to believe. I definitely want to look into that more in the near future. As you say, I'm sure it will be a Big Deal for some users.

For posterity, the issue is configuring otel-collector to listen on an `https` connection.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        tags: Attributes = None,
+    ) -> None:
+        """Gauge stat."""
+        # To be implemented
```
nit: maybe raise the `NotImplementedError` here, and below, so it's clear users should not use this.

ferruzzi replied:
Gah, I did consider that and didn't.  I'll add that.

ferruzzi replied:
Actually, I did.   It'll keep throwing exceptions while I work on this if I do that.  The way it is done here, it just carries on every time it tries to emit a metric type I have not yet implemented.

I could do a warnings.warn() here and in the timers, if you'd like though??

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+
+
+class MetricsMap:
+    """Stores Otel Counters."""
```
Nit: more context/info?

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        Deletes a counter.
+
+        :param name: The name of the counter to fetch or create.
```
copy/paste, you're not creating any here.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,234 @@
+
+
+def get_otel_logger(cls) -> SafeOtelLogger:
+    """Get Otel logger"""
```
nit: I'd add more to this or just drop.

o-nikolas commented on airflow/metrics/otel_logger.py:
I'm not seeing any tests for this module or the classes in it? Do we think there'd be any value in that?

ferruzzi replied:
Yup, I will definitely add a test suite.

ferruzzi replied:
Added in https://github.com/apache/airflow/pull/30873/commits/f794da05f71dc4f5042e40aeb2b7b86f886e206e

ferruzzi committed:
PR Fixes Round 1

- Converts a couple of stray prints to logs
- Attempts to consolidate the redundant metrics by adding some logic in incr() and decr() to emit both versions of the metric name in one call
- Removes some pointless docstrings
- Removes the accidental inclusion of BaseInstrument

ferruzzi committed:
Implement rate as a "percentage" of the time to emit the metric, same as StatsD.

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,242 @@
+        :param stat: The name of the stat to increment.
+        :param count: A positive integer to add to the current value of stat.
+        :param rate: Value between 0 and 1 representing the "percentage" of
+            the time to emit the metric where 1 = 100%.
```
Still trying to think of a better way to phrase this, open to suggestions.

ferruzzi replied:
Howard replied below 

> Why not add ‘sampling rate’ instead of just ‘percentage?’ I think it might be more concise if we could describe it as “value between 0 and 1 that represents the sampled rate at which the metric is going to be emitted.”

I like that.  I'll make the change.

o-nikolas replied:
IMHO this is more of a  probability, since a random dice roll is involved. If you set it to 0.5 and then call it 10 times you're not guaranteed to get 5 of them emitted. You may get 6 emitted or 4 or even all 10 if you're (un)lucky. A sample rate to me would be more deterministic, but you could also just stitch both bits of verbiage together and call it a probabilistic sample rate :grin:

ferruzzi committed:
Add unit tests

Also:
- Make some adjustments based on the unit test results
- Rephrase the docstring for the `rate` parameter

ferruzzi committed:
fix logging string formats

o-nikolas commented on tests/core/test_otel_logger.py:
```diff
@@ -0,0 +1,156 @@
+        assert not _is_up_down_counter("this_is_not_a_udc")
+
+    def test_stat_name_must_be_a_string(self):
+        self.stats.incr(42)
```
Shouldn't this and the other incorrect name cases throw an exception?

ferruzzi replied:
Interesting... it isn't caught as an exception because it just gets folded into an f-string [here](https://github.com/apache/airflow/pull/30873/files#diff-bc07cf67d2d211678aaefd4d99b1e8c6e2c6e26a6987e85c5397ccc78fed47feR89) but the incr() parameter is type-hinted as a string.  I could throw an `if not isInstance(name, str): raise ..." but I wonder if a "validate_name()" helper method that asserts it is a string and truncates as needed might be a better answer?

ferruzzi replied:
Thinking on this more, SHOULD they?  Do we really want Airflow blowing up because someone implemented a metric with a bad name?  Maybe this should be logged as a warning?

I'll add a more thorough name screening/validation, but maybe an exception isn't the right failure behavior here.

ferruzzi replied:
Also of note is that the name MAY contain digits, the [OTel spec](https://opentelemetry.io/docs/reference/specification/metrics/api/#instrument-name-syntax) says it must START with an alpha and subsequent characters must belong to the alphanumeric characters, ‘_’, ‘.’, and ‘-’.   Since we are prefixing `airflow.` to the names, then technically the name "42" (as a string) here would qualify as a legal name since it is later getting folded into an f-string, which confuses things.   I will run the validation checks assuming there is no prefix in case we decide to drop that at some point for some reason.

howardyoo replied:
> Thinking on this more, SHOULD they? Do we really want Airflow blowing up because someone implemented a metric with a bad name? Maybe this should be logged as a warning?
> 
> I'll add a more thorough name screening/validation, but maybe an exception isn't the right failure behavior here.

We definitely do not want Airflow to blow up, simply because of a bad metric naming. What we should do is to block those `illegal` metric names from getting to the metric registry, and simply posting warning on the airflow log to state that the following metric `XYZ` is illegal (does not start with alphabet) and therefore it was not included.

howardyoo replied:
> 
I do believe, however, that the default prefix `airflow` can be changed, is it not? In that case, one can set the prefix to `42` and that might break the OTEL spec..?

ferruzzi replied:
What I have now (tests are running locally, will push this evening) is a check which validates the name by itself, validates the optional prefix if it's included, and then checks the length of the combined `prefix.name`.  So it should cover us.  Also added unit tests for each of those.

o-nikolas commented on tests/core/test_otel_logger.py:
```diff
@@ -0,0 +1,156 @@
+        self.meter.get_meter().create_counter.assert_has_calls(expected_calls, any_order=True)
+        assert self.meter.get_meter().create_counter.call_count == len(expected_calls)
+
+    @mock.patch("random.random", side_effect=[0.1, 0.9])
```
Nice use of multiple side effects :+1:

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,261 @@
+        :param rate: value between 0 and 1 that represents the sampled rate at
+            which the metric is going to be emitted.
+        :param tags: Tags to append to the stat.
+        :param back_compat_name: If possible, the metric will also be emitted
```
When would it not be possible to use the back compat name?

ferruzzi replied:
I think that needs to be rephrased,  It might possibly not pass the validator, and it may also get truncated.

ferruzzi replied:
This parameter has been reverted/removed, so no longer relevant.  Resolving.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,261 @@
+        if rate < 1 and random.random() > rate:
+            return
+
+        if back_compat_name and self.metrics_validator.test(back_compat_name):
```
Does the metrics_validator do the assertion of the name length for otel's short naming? Otherwise it's going to blow up when it hits the hotel code.

ferruzzi replied:
No, I think I need to truncate it on the following line(89).  I'll also add a unit test which passes in a back_compat_name that is too long and makes sure it doesn't blow up.

ferruzzi replied:
This parameter has been reverted/removed, so no longer relevant. Resolving.

howardyoo reviewed (commented):
commented

ferruzzi committed:
Move Otel debug exporter behind a config flag

ferruzzi committed:
Remove the back_compat_name parameter from incr() and decr()

ferruzzi committed:
Fix name validation and add more unit tests

ferruzzi committed:
Add context to UpDownCounter comment

ferruzzi committed:
Move opentelemetry-api from the otel extras into install_requires

ferruzzi committed:
Move opentelemetry-exporter-otlp from the otel extras into install_requires

potiuk commented on airflow/config_templates/config.yml:
```diff
@@ -1019,6 +1019,13 @@ metrics:
+    otel_debugging_on:
+      description: |
+        If True, all metrics are also emitted to the console.  Defaults to False.
+      version_added: 2.5.1
```
```suggestion
      version_added: 2.7.0
```

ferruzzi committed:
fix version_added

ferruzzi committed:
Added a couple more test cases

ferruzzi committed:
Revise metrics_map key generation method and update unit tests accordingly.

potiuk commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+    https://opentelemetry.io/docs/reference/specification/metrics/api/#instrument-name-syntax
+    """
+
+    def test(value) -> bool:
```
NIT: maybe `_test_valid_name` ? (I try to avoid such short, generic names, that could be easily mistaken with some import, also `_` would indicate it's just internal.

ferruzzi replied:
Yup, that's fair.  I was thinking it was nested inside another method so it would be safe, but I shouldn't make that kind of assumption, I guess.

potiuk commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+        """Creates a new counter or up_down_counter for the provided name."""
+        otel_safe_name = name[:OTEL_NAME_MAX_LENGTH]
+        if name != otel_safe_name:
+            warnings.warn(
```
Should we simply raise error in this case? The problem with this warning is that user who will see it will not be able to do anything with it, and those are the worst kinds of warnings. This check should be targeted not for the users of airflow, but for those who create new metrics, and for them, just failing it here would be IMHO much betters because it will prevent them from adding the wrong metrics name in the first place.

ferruzzi replied:
Until/unless we deprecate all of the old names that are too long, I think this needs to stay.  I can add a TODO comment noting that it can be changed to an exception if the old names are removed, but there are existing names which would start breaking if we made this an exception.

o-nikolas replied:
I agree, it would be awesome to strictly prohibit names that break OTEL. I think a middle ground can be struck here:
maintain a list of the existing metric names that are too long, and warn for those, anything else/new throw an exception. Then we can work to deprecate the "grandfathered in" long metric names overtime.

ferruzzi replied:
We did discuss this at the beginning of his project in the terms of a "metrics registry". The problem with having a list of names which are too long is that most (all?) of the names which would cause an issue are dynamically generated.  

Some are easy to catch with a "startswith()" like  [f"local_task_job.task_exit.{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}"](https://github.com/apache/airflow/blob/main/airflow/jobs/local_task_job_runner.py#L298), but then we would also have [f"dagrun.{dag.dag_id}.first_task_scheduling_delay"](https://github.com/apache/airflow/blob/main/airflow/models/dagrun.py#L929) and [<job_name>_heartbeat_failure](https://github.com/apache/airflow/blob/main/airflow/jobs/job.py#L215) to figure out how to match.  

I guess we could use some regex matching shenanigans, if we want t do this.

This also goes back to the early discussions:  Is OTel going to be one available option for metrics alongside StatsD, or is it going to be the only option?   If it is going to be a pluggable option as we decided, then I don't feel like enforcing its naming policies universally is a great idea.

o-nikolas replied:
> Some are easy to catch with a "startswith()" like f"local_task_job.task_exit.{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}", but then we would also have f"dagrun.{dag.dag_id}.first_task_scheduling_delay" and <job_name>_heartbeat_failure to figure out how to match.
>
> I guess we could use some regex matching shenanigans, if we want t do this.

Yeah I think a regex pattern that you register for each one would be a nice solution.

> If it is going to be a pluggable option as we decided, then I don't feel like enforcing its naming policies universally is a great idea.

I don't think it's unreasonable to enforce the minimum name length across all possible metrics platforms?

ferruzzi replied:
I've added a `set` which has all of the currently potentially-too-long names as regex patterns and added a check against that.  If the name is too long and not on that exemption list then an exception is thrown.  Let me know hat you two thin of the way I implemented it.  This should let us deprecate the old names over time and prevent anyone from adding new metrics that would break the standard.

potiuk commented on setup.cfg:
```diff
@@ -117,6 +117,9 @@ install_requires =
     marshmallow-oneofschema>=2.0.1
     mdit-py-plugins>=0.3.0
+    # Pip can not find a version that satisfies constraints if opentelemetry-api is not pinned.
+    opentelemetry-api==1.15.0
```
I believe you should be able to replace it with`>= 1.15.0` now. The problems were likely in Python 3.7 only and caused by importlib_metada incompatibilities. Those will be finally solved when Celery releases 5.3 (they have rc1 as of 11 May) but I think resolution should work.

https://pypi.org/project/celery/#history

ferruzzi replied:
Sweet, I'll test it and see if it works now.  I had missed that update, thanks.

ferruzzi replied:
Seems to still run into the issue when I use `>=`.  `breeze ci-image build --upgrade-to-newer-dependencies --python 3.8` is still running after 2500s

potiuk reviewed (approved):
Few NITs - I see that OTEL is not a "primary" interface for metrics generation, and if you think that is simpler and better from our perspective, I am also fine with it, but we will likely have to figure a way to modify our "Stats" interface later to add the spans that will be needed to be added to the otel-emitted metrics. But for now I think it will be good to get merged.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+            raise InvalidStatsNameException
+        test(name)
+        if prefix:
+            test(prefix)
```
Is there any value to testing the prefix by itself? Is there a time where it won't be prefixed/appended to the name?

ferruzzi replied:
In fact, now that you pointed it out, `prefix` isn't an optional arg in this final version, so that can be refactored completely.  I'll make the change.

ferruzzi replied:
The reason I did it that way was https://github.com/apache/airflow/pull/30873#discussion_r1181957232 but I can rework this smarter.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+            counter = self.meter.create_counter(name=otel_safe_name)
+
+        counter_type = str(type(counter)).split(".")[-1][:-2]
+        logging.debug("--> created %s as type: %s", otel_safe_name, counter_type)
```
Nit: Why the ascii arrow?

ferruzzi replied:
Personal debugging that I forgot to trim out.   Thanks for catching it.

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+
+    def get_counter(self, name: str, attributes: Attributes = None):
+        """
+        Returns the counter and creates a new one if it does not exist.
```
Just a nit-suggestion, feel free to take it or not :) 
```suggestion
        Returns the counter or creates a new one if it does not exist.
```

ferruzzi replied:
Interesting.  That is pretty awkward ether way, isn't it.   If the counter didn't exist then it creates it and returns it.  So either way it is returning the counter.  A semicolon would maybe make it less ambiguous?

`Returns the counter; creates a new one if it did not exist.` maybe?

o-nikolas replied:
Yeah, I'd agree with semicolon :+1:

o-nikolas commented on airflow/metrics/otel_logger.py:
```diff
@@ -0,0 +1,289 @@
+    # TODO:  figure out https instead of http ??
+    endpoint = f"http://{host}:{port}/v1/metrics"
+
+    logging.info("[Metric Exporter] Connecting to OTLP at ---> %s", endpoint)
```
Nit: Expand the acronym perhaps? I at least wouldn't know what that stood for.

ferruzzi replied:
Fair callout.   It's a little awkward.  OTLP is the "acronym" for Open Telemetry Protocol.  The "L" is.. silent... I guess?   :shrug:    I'll replace it with a more accurate statement: "Connecting to Open Telemetry Collector at", I suppose since `endpoint` is the URI for the Collector's "input" uri/port.

o-nikolas replied:
Sounds good!

ferruzzi committed:
PR Fixes Round 2

ferruzzi committed:
docstring tweaks

ferruzzi committed:
Adds long metric name exemption checks and related unit tests

ferruzzi committed:
Drop the Java-style "Generic"

ferruzzi committed:
Adds a test to ensure the exemption list does not grow and a comment about why.
